azkaban-common/src/main/java/azkaban/project/FlowTrigger.java

```diff
@@ -0,0 +1,108 @@
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package azkaban.project;
+
+import azkaban.utils.Props;
+import com.google.common.base.Preconditions;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * FlowTrigger is an immutable class
+ * which contains a list of FlowTriggerDependency and a schedule.
+ */
+public class FlowTrigger {
+
+  private final List<FlowTriggerDependency> dependencies;
+  private final FlowTriggerSchedule schedule;
+
+  private FlowTrigger(final FlowTriggerSchedule schedule,
+      final List<FlowTriggerDependency> dependencies) {
+    validateDependencies(dependencies);
+    this.schedule = schedule;
+    this.dependencies = Collections.unmodifiableList(dependencies);
+  }
+
+
+  /**
+   * check uniqueness of dependency.name
+   */
+  private void validateDepNameUniqueness(final List<FlowTriggerDependency> dependencies) {
+    final Set<String> seen = new HashSet<>();
+    for (final FlowTriggerDependency dep : dependencies) {
+      // set.add() returns false when there exists duplicate
+      Preconditions.checkArgument(seen.add(dep.getName()), String.format("duplicate dependency"
+          + ".name %s found, dependency.name should be unique", dep.getName()));
+    }
+  }
+
+  /**
+   * check uniqueness of dependency type and params
+   */
+  private void validateDepDefinitionUniqueness(final List<FlowTriggerDependency> dependencies) {
+    final Set<String> seen = new HashSet<>();
+    for (final FlowTriggerDependency dep : dependencies) {
+      final Props props = dep.getPropsCopy();
+      // set.add() returns false when there exists duplicate
+      Preconditions.checkArgument(seen.add(dep.getType() + ":" + props.toString()), String.format
+          ("duplicate "
+              + "dependency"
+              + "config %s found, dependency config should be unique", dep.getName()));
+    }
+  }
+
+  private void validateDependencies(final List<FlowTriggerDependency> dependencies) {
+    Preconditions.checkNotNull(dependencies);
+    validateDepNameUniqueness(dependencies);
+    validateDepDefinitionUniqueness(dependencies);
+  }
+
+  @Override
+  public String toString() {
+    return "FlowTrigger{" +
+        "dependencies=" + this.dependencies +
+        ", schedule=" + this.schedule +
+        '}';
+  }
+
+  public List<FlowTriggerDependency> getDependencies() {
+    return this.dependencies;
+  }
+
+  public static class FlowTriggerBuilder {
```

Comment:
Is a builder necessary?

Reply:
it's in grey area. I think either way is fine.